Document the use of lemmas (#2709)
ucsd-progsys/liquidhaskell: docs/mkDocs/docs/specifications.md

           -> (Ix Val t1 -> Ix Val t2)
           -> Ix Val (TFun t1 t2) @-}
  VFun :: Ty -> Ty -> (Val -> Val) -> Val
```

## Using Lemmas

Sometimes additional context information is need for LH to verify a specification.
To accomplish this, we can use [lemmas](https://en.wikipedia.org/wiki/Lemma_(mathematics)).

The `Language.Haskell.Liquid.ProofCombinators` module provides the `?` combinator,
which is an infix variant of `const` optimized to pass the refined type of its
second argument to the constraint context of the first argument.

For example, `e ? lemma` lets you use the type of the expression `lemma`
when checking that expression `e` has some expected refinement type.

Schematically, `?` allows to use the `q x` to check that `e` satisfies `p x e`
in the following function.

```haskell
f :: x:t0 -> {v:t1 | p x v}
f x = e ? lemma x

lemma :: y:t0 -> { q y }
lemma = ...
```

> Dead bindings inside a `where` clause or `let` expression can be used

Reviewer: ```suggestion
Bindings inside a `where` clause or `let` expression can be used
```

to accomplish this as well:

```haskell
f :: x:t0 -> {v:t1 | p x v}
f x = e
  where

>     lemma0 = lemma x -- dead binding

Reviewer: ```suggestion
    -- the lemma0 binding is unused by e but it is still available
    -- for Liquid Haskell to use it
    lemma0 = lemma x
```

-- alternatively

f :: x:t0 -> {v:t1 | p x v}

> f x = let lemma x {- dead binding -} in e

Reviewer: ```suggestion
f x =
  let -- the lemma0 binding is unused by e but it is still available
      -- for Liquid Haskell to use it
      lemma0 = lemma x
   in e
```

```

There are some differences to both approaches:

- Scope: while the former scopes only over `e`,
  the latter inserts the type in all constraints arising from locations

>   where the binder is in scope.

Reviewer: `(f e) ? lemma` only inserts the lemma type in subtyping constraints at `f e`.

If a subtyping constraint at `e` needs the lemma too, you would need to write `(f (e ? lemma)) ? lemma`.

in the following function.

```haskell
f :: x:t0 -> {v:t1 | p x v}
f x = e ? lemma x

lemma :: y:t0 -> { q y }
lemma = ...
```

Bindings inside a `where` clause or `let` expression can be used
to accomplish this as well:

```haskell
f :: x:t0 -> {v:t1 | p x v}
f x = e
  where
    -- the lemma0 binding is unused by e, but it is still available
    -- for Liquid Haskell to use it
    lemma0 = lemma x

-- alternatively

f :: x:t0 -> {v:t1 | p x v}
f x =
  let -- the lemma0 binding is unused by e, but it is still available
      -- for Liquid Haskell to use it
      lemma0 = lemma x
   in e
```

There are some differences to both approaches:

- Scope: The former inserts the lemma type in the subtyping constraints of
  the expression value (`e` in the examples above) only, for example:
  `(f e) ? lemma` inserts the `lemma` type at `f e`, but if a subtyping
  constraint at `e` needs the `lemma` too, you would need to write
  `(f (e ? lemma)) ? lemma`.
  In comparison, the latter option is blunter: it inserts the type in all
  locations where the binder is in scope, so a single binding inserts it into
  both `e` and `f e` in this example.

> - The dead binding approach also has the advantage that the lemmas appear

Reviewer: ```suggestion
- The binding approach also has the advantage that the lemmas appear
```
